firewall/master.go

```diff
@@ -31,91 +31,170 @@ import (
+		conn.Verdict = network.VerdictAccept
 		return
 	}
 
-	// grant self
-	if comm.Process().Pid == os.Getpid() {
-		log.Infof("firewall: granting own communication %s", comm)
-		comm.Accept("")
-		return
+	// check if process is communicating with itself
+	if pkt != nil {
+		if conn.Process().Pid >= 0 && pkt.Info().Src.Equal(pkt.Info().Dst) {
+			// get PID
+			otherPid, _, err := process.GetPidByEndpoints(
+				pkt.Info().RemoteIP(),
+				pkt.Info().RemotePort(),
+				pkt.Info().LocalIP(),
+				pkt.Info().LocalPort(),
+				pkt.Info().Protocol,
+			)
+			if err == nil {
+
+				// get primary process
+				otherProcess, err := process.GetOrFindPrimaryProcess(pkt.Ctx(), otherPid)
+				if err == nil {
+
+					if otherProcess.Pid == conn.Process().Pid {
+						conn.Accept("connection to self")
+						return
+					}
+				}
+			}
+		}
 	}
 
 	// get profile
-	p := comm.Process().Profile()
-
-	// check for any network access
-	if p.BlockScopeInternet() && p.BlockScopeLAN() {
-		log.Infof("firewall: denying communication %s, accessing Internet or LAN not permitted", comm)
-		comm.Deny("accessing Internet or LAN not permitted")
+	p := conn.Process().Profile()
+	if p == nil {
+		conn.Block("no profile")
 		return
 	}
-	// continueing with access to either Internet or LAN
 
-	// check endpoint list
-	// FIXME: comm.Entity.Lock()
-	result, reason := p.MatchEndpoint(comm.Entity)
-	// FIXME: comm.Entity.Unlock()
+	// check conn type
+	switch conn.Scope {
+	case network.IncomingHost, network.IncomingLAN, network.IncomingInternet, network.IncomingInvalid:
+		if p.BlockInbound() {
+			if conn.Scope == network.IncomingHost {
+				conn.Block("inbound connections blocked")
+			} else {
+				conn.Deny("inbound connections blocked")
+			}
+			return
+		}
+	case network.PeerLAN, network.PeerInternet, network.PeerInvalid:
+		// Important: PeerHost is and should be missing!
+		if p.BlockP2P() {
+			conn.Block("direct connections (P2P) blocked")
+			return
+		}
+	}
+
+	// check scopes
+	if conn.Entity.IP != nil {
+		classification := netutils.ClassifyIP(conn.Entity.IP)
+
+		switch classification {
+		case netutils.Global, netutils.GlobalMulticast:
+			if p.BlockScopeInternet() {
+				conn.Deny("Internet access blocked") // Block Outbound / Drop Inbound
+				return
+			}
+		case netutils.SiteLocal, netutils.LinkLocal, netutils.LocalMulticast:
+			if p.BlockScopeLAN() {
+				conn.Block("LAN access blocked") // Block Outbound / Drop Inbound
+				return
+			}
+		case netutils.HostLocal:
+			if p.BlockScopeLocal() {
+				conn.Block("Localhost access blocked") // Block Outbound / Drop Inbound
+				return
+			}
+		default: // netutils.Invalid
+			conn.Deny("invalid IP") // Block Outbound / Drop Inbound
+			return
+		}
+	} else if conn.Entity.Domain != "" {
+		// DNS Query
+		// DNS is expected to resolve to LAN or Internet addresses
+		// TODO: handle domains mapped to localhost
+		if p.BlockScopeInternet() && p.BlockScopeLAN() {
```

Comment:
Do we only block the connection if both, `ScopeInternet` **and** `ScopeLAN are marked for blocking?

Reply:
For DNS Queries, yes, as we don't know in which scopes the IPs of the response will be. Additional filtering is done by the `FilterDNSResponse` fn, which then looks at these separately.